File: scalechain-proto-codec/src/test/scala/io/scalechain/blockchain/proto/codec/BitcoinProtocolCodecSpec.scala
package io.scalechain.blockchain.proto.codec

/**
  * Created by kangmo on 2/7/16.
  */
class BitcoinProtocolCodecSpec {

}


package io.scalechain.blockchain.cli

import io.scalechain.blockchain.ErrorCode
import io.scalechain.blockchain.ProtocolCodecException
import io.scalechain.blockchain.cli.BlockDirectoryReader
import io.scalechain.blockchain.cli.BlockReadListener
import io.scalechain.blockchain.proto.Block
import io.scalechain.blockchain.{ErrorCode, ProtocolCodecException}
import io.scalechain.blockchain.proto.Block
import org.scalatest._

/**
  * Created by kangmo on 11/2/15.
  */
class BitcoinProtocolCodecSpec extends FlatSpec with BeforeAndAfterEach with ShouldMatchers {
  this: Suite =>

  override def beforeEach() {
    // set-up code
    //

    super.beforeEach()
  }

  override def afterEach() {
    super.afterEach()
    // tear-down code
    //
  }

  "encode" should "successfully encode a message" in {
  }

  "decode" should "return an incomplete message with an empty message vector" in {
  }

  "decode" should "return an incomplete message with a non-empty message vector" in {
  }

  "decode" should "return no incomplete message with a non-empty message vector" in {
  }
}

File: scalechain-proto-codec/src/test/kotlin/io/scalechain/blockchain/proto/codec/BitcoinProtocolCodecSpec.kt
package io.scalechain.blockchain.proto.codec

import io.kotlintest.KTestJUnitRunner
import io.kotlintest.matchers.Matchers
import io.kotlintest.specs.FlatSpec
import io.netty.buffer.ByteBuf
import io.netty.buffer.Unpooled
import io.scalechain.blockchain.proto.ProtocolMessage
import io.scalechain.blockchain.proto.SendHeaders
import io.scalechain.blockchain.proto.Verack
import io.scalechain.blockchain.proto.Version
import org.junit.runner.RunWith

/**
  * Created by kangmo on 11/2/15.
  */
@RunWith(KTestJUnitRunner::class)
class BitcoinProtocolCodecSpec : FlatSpec(), Matchers {
  val codec = BitcoinProtocolCodec(BitcoinProtocol())

  init {
    "encode" should "successfully encode a message that can be decodable" {
      val messageList = mutableListOf<Any>()
      val buffer1 = Unpooled.buffer()
      val buffer2 = Unpooled.buffer()
      // We already have tests for each class that extends ProtocolMessage
      // Test if BitcoinProtocolCodec.encode works well for two protocol messages.
      codec.encode(Verack(), buffer1)
      codec.decode(buffer1, messageList)
      messageList shouldBe listOf(Verack())

      codec.encode(SendHeaders(), buffer2)
      codec.decode(buffer2, messageList)
      messageList shouldBe listOf(Verack(), SendHeaders())
    }

    "decode" should "return a message without leaving any input bytes" {
      val messageList = mutableListOf<Any>()
      val buffer = Unpooled.buffer()

      codec.encode(Verack(), buffer)
      codec.decode(buffer, messageList)
      messageList shouldBe listOf(Verack())
      buffer.readableBytes() shouldBe 0
    }

    "decode" should "return two messages without leaving any input bytes" {
      val messageList = mutableListOf<Any>()
      val buffer = Unpooled.buffer()

      codec.encode(Verack(), buffer)
      codec.encode(SendHeaders(), buffer)

      codec.decode(buffer, messageList)
      messageList shouldBe listOf(Verack(), SendHeaders())
      buffer.readableBytes() shouldBe 0
    }

    "decode" should "return a message leaving incomplete input bytes" {
      val messageList = mutableListOf<Any>()
      val buffer = Unpooled.buffer()

      codec.encode(Verack(), buffer)
      buffer.writeByte(100) // Write an additional byte

      codec.decode(buffer, messageList)
      messageList shouldBe listOf(Verack())
      buffer.readableBytes() shouldBe 1
      buffer.readByte()      shouldBe 100.toByte()
    }

    "decode" should "return two messages leaving incomplete input bytes" {
      val messageList = mutableListOf<Any>()
      val buffer = Unpooled.buffer()

      codec.encode(Verack(), buffer)
      codec.encode(SendHeaders(), buffer)
      buffer.writeByte(100) // Write an additional byte

      codec.decode(buffer, messageList)
      messageList shouldBe listOf(Verack(), SendHeaders())
      buffer.readableBytes() shouldBe 1
      buffer.readByte()      shouldBe 100.toByte()
    }
  }
}
